src/inet/TCPEndPoint.cpp

```diff
@@ -696,6 +696,67 @@ INET_ERROR TCPEndPoint::GetLocalInfo(IPAddress * retAddr, uint16_t * retPort)
     return res;
 }
 
+INET_ERROR TCPEndPoint::GetInterfaceId(InterfaceId * retInterface)
+{
+    if (!IsConnected())
+        return INET_ERROR_INCORRECT_STATE;
+
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    // TODO: Does netif_get_by_index(mTCP->netif_idx) do the right thing?  I
+    // can't quite tell whether LwIP supports a specific interface id for TCP at
+    // all.  For now just claim no particular interface id.
+    *retInterface = INET_NULL_INTERFACEID;
+    return INET_NO_ERROR;
+#endif // CHIP_SYSTEM_CONFIG_USE_LWIP
+
+#if CHIP_SYSTEM_CONFIG_USE_SOCKETS
+    union
+    {
+        sockaddr any;
+        sockaddr_in6 in6;
+#if INET_CONFIG_ENABLE_IPV4
+        sockaddr_in in;
+#endif // INET_CONFIG_ENABLE_IPV4
+    } sa;
+
+    memset(&sa, 0, sizeof(sa));
+    socklen_t saLen = sizeof(sa);
+
+    if (getpeername(mSocket, &sa.any, &saLen) != 0)
+    {
+        return chip::System::MapErrorPOSIX(errno);
+    }
+
+    if (sa.any.sa_family == AF_INET6)
+    {
+        if (IPAddress::FromIPv6(sa.in6.sin6_addr).IsIPv6LinkLocal())
+        {
+            *retInterface = sa.in6.sin6_scope_id;
+        }
+        else
+        {
+            // TODO: Is there still a meaningful interface id in this case?
+            *retInterface = INET_NULL_INTERFACEID;
```

Comment:
As far as I know, sin6_scope_id will only be an interface ID for link local addresses.  So in short, no, I don't believe an interface ID can be extracted for addresses other than link-local.  And I can't think of a reason why we would need this for non-link-local addresses.  In theory we could bind to a specific interface at which we know a device may reach us.  But I don't know why we would want to or need to.

As to the other possible overloaded meanings of sin6_scope_id (zones or site indices?), it's hard to imagine these would be useful to us, or could even be dealt with in a portable way.  I don't know much about those though, except that they are not interface indices.

TLDR: It looks to me like the code is doing the correct thing here.

Reply:
Thank you for looking into this!  This part I was really not sure about, as you can tell from the comments.  ;)